### ISSUE TITLE
C# Support for SQLCLR Host

### ISSUE THREAD

**Issue by tastyeggs** [google/protobuf#3586]
I'm trying to deploy the Protobuf.dll library, and generated classes for my models into SQL Server using SQLCLR Assembly feature, so that I can decode protobuf-encoded columns as part of a SQL stored proc.

I'd like to refrain from using the UNSAFE permission set, so that this could be deployed to PaaS services such as SQL Azure.

Compatibility is easy to achieve (have it working for myself, including deployment to SQL Server):
1. Make generated `private readonly pbr::FileDescriptor descriptor` fields static
2. Update generated WellKnownTypes, so this is reflected in all generated C# classes within the Protobuf library
3. Replace all occurrences of HashSet<T> with IDictionary<T,bool> (not too many), since HashSet is decorated with `[HostProtectionAttribute(SecurityAction.LinkDemand, MayLeakOnAbort = true)]`. SQLCLR .NET Host does not allow MayLeakOnAbort. `ContainsKey` and `Add`, which are the primary operations being used, have very similar performance characteristics to HashSet, so there is no drawback here.

Happy to submit a Pull Request, but before I put in the effort to figure out the entire build system and how to run tests etc, want to throw this idea out there in case anyone has thoughts / objections.

**Comment by jskeet** [comment]
Will try to look at this tomorrow, but it may take a bit of time.

**Comment by jskeet** [comment]
I don't follow the part about point 1 - if you mean fields like [this](https://github.com/google/protobuf/blob/6699f2cf64c656d96f4d6f93fa9563faf02e94b4/csharp/src/Google.Protobuf/WellKnownTypes/Duration.cs#L20), they're already static. Could you give an example of a field you'd want to be static that isn't?

**Comment by tastyeggs** [comment]
I apologize, I meant static fields need to be marked readonly. I've edited the issue above.

**Comment by jskeet** [comment]
Ah, right - yup, that's fine. Using Dictionary instead of HashSet sounds fine to me too.

If you're happy to create the pull request, that would be great! I'd suggest separating it into three commits if possible:

- Codegen change
- The result of running the codegen (test/sample protos and well-known types)
- The HashSet change

Let me know if you have any problems.

**Comment by pherl** [comment]
We are closing open issues. Feel free to send PRs and reopen if you'd like to follow up on the issue.

**Comment by chriscosta** [comment]
Was this issue released? i am in need of a protobuf library that can be used against sql server on a linux based machine which currently does not support unsafe libraries

**Comment by jskeet** [comment]
@chriscosta: I don't believe the change was ever made. There's certainly still code using `HashSet<T>`.

**Comment by chriscosta** [comment]
> @chriscosta: I don't believe the change was ever made. There's certainly still code using `HashSet<T>`.

I was hoping there was! Is there any workaround for this? @tastyeggs , u say that u have implemented it for you! Any chance you could share what you have done?

**Comment by tastyeggs** [comment]
@chriscosta 

We didn't go all the way with this approach, however verified it as working in a proof-of-concept. However I can get you half the way here, where (1) and (2) above are addressed (they're the same problem, where a static variable isn't marked as readonly, something SQLCLR doesn't like). (3) shouldn't be too difficult to do, just find and replace all instances and usages of HashSet<T> with Dictionary<T, bool>.

What we have is a bit kludgy, since we didn't modify the protoc tool.

1. A dotnet msbuild task that adds in "readonly" to `pbr::FileDescriptor` on generated .cs files, something like this.
```
<Target Name="GenerateCsForProtoC" Inputs="@(ProtoC)" Outputs="$(MSBuildProjectDirectory)\obj\$(Configuration)\ProtoC-v1\$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__'))%(ProtoC.Filename).cs" BeforeTargets="BeforeBuild;BeforeCompile">
    <MakeDir Directories="$(MSBuildProjectDirectory)/obj/$(Configuration)/ProtoC-v1/$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__'))" />
    <ReadLinesFromFile File="$(SolutionDir)/Solution Items/ProtoBuf/ProtoCReplPsCmd.txt">
        <Output TaskParameter="Lines" PropertyName="PsCmd" />
    </ReadLinesFromFile>
    <Exec Command="&quot;$(NuGetPackageRoot)/Google.Protobuf.Tools/3.4.0/tools/windows_x64/protoc.exe&quot; --proto_path=&quot;$(SolutionDir)\Libraries\Egg.Grocery.Shared&quot; --proto_path=&quot;$(SolutionDir)\Libraries\Egg.Grocery.Shared.Backend&quot;  --proto_path=&quot;$(MSBuildProjectDirectory)&quot; --csharp_out=&quot;$(MSBuildProjectDirectory)/obj/$(Configuration)/ProtoC-v1/$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__').TrimEnd('\\'))&quot; &quot;%(ProtoC.FullPath)&quot; 2>&amp;1 | powershell -Command &quot;$(PsCmd)&quot;" IgnoreStandardErrorWarningFormat="false" />
    <PostProcessGeneratedCsFile InputFilename="$(MSBuildProjectDirectory)\obj\$(Configuration)\ProtoC-v1\$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__'))%(ProtoC.Filename).cs"
        OutputFilename="$(MSBuildProjectDirectory)\obj\$(Configuration)\ProtoC-v1\$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__'))%(ProtoC.Filename).cs"
        MatchText="private static pbr::FileDescriptor"
        ReplacementText="private static readonly pbr::FileDescriptor" />
    <ItemGroup>
        <Compile Include="$(MSBuildProjectDirectory)\obj\$(Configuration)\ProtoC-v1\$([System.String]::Copy('%(ProtoC.RelativeDir)').Replace(':','__DRIVE__'))%(ProtoC.Filename).cs">
            <DependentUpon>%(ProtoC.FullPath)</DependentUpon>
        </Compile>
    </ItemGroup>
</Target>
```

2.  A change to the Protobuf C# nuget library, so this readonly modifier is added to WellKnownType IMessages:, I pushed our code here: https://github.com/tastyeggs/protobuf/commit/4395ffc1861feff9f3f5be1617fe38504a2c3647.

You can avoid the kludgy the BeforeCompile msbuild task by modifying the protoc compiler, in particular this file: `src\google\protobuf\compiler\csharp\csharp_reflection_class.cc`, and add in the readonly modifier.

3. If you want to execute the serialize/deserialize functionality within SQLCLR, you'll need to replace HashSet instances with IDictionary, but since we aren't using this in production, we haven't done this.

Hope this help.

**Comment by chriscosta** [comment]
Thank you so much for taking the time to explain what needs to be done. I am not sure if i will be able to do all this, i will give it a go, unless my boss decides to dump the whole linux os and go with windows instead.
